src/OpenAIGym.jl

```diff
@@ -91,22 +99,68 @@ function actionset(A::PyObject)
     end
 end
 
-
 function Reinforce.actions(env::AbstractGymEnv, s′)
     actionset(env.pyenv[:action_space])
 end
 
 pyaction(a::Vector) = Any[pyaction(ai) for ai=a]
 pyaction(a) = a
 
-function Reinforce.step!(env::GymEnv, a)
-    # info("Going to take action: $a")
+"""
+`reset!` for PyArray state types
+"""
+function Reinforce.reset!(env::GymEnv{T}) where T <: PyArray
+    env.state = PyArray(pycall!(env.pystate, env.pyreset, PyObject))
+    return gymreset!(env)
+end
+
+"""
+`reset!` for non PyArray state types
+"""
+function Reinforce.reset!(env::GymEnv{T}) where T
+    pycall!(env.pystate, env.pyreset, PyObject)
+    env.state = convert(T, env.pystate)
+    return gymreset!(env)
+end
+
+function gymreset!(env::GymEnv{T}) where T
+    env.reward = 0.0
+    env.total_reward = 0.0
+    env.actions = actions(env, nothing)
+    env.done = false
+    return env.state
+end
+
+"""
+`step!` for PyArray state
+"""
+function Reinforce.step!(env::GymEnv{T}, a) where T <: PyArray
     pyact = pyaction(a)
-    s′, r, env.done, env.info = env.pyenv[:step](pyact)
-    env.reward = r
+    pycall!(env.pystepres, env.pystep, PyObject, pyact)
+
+    env.pystate, r, env.done, env.info =
+        convert(Tuple{PyObject, Float64, Bool, PyObject}, env.pystepres)
+
+    env.state = PyArray(env.pystate)
+
+    env.total_reward += r
+    return (r, env.state)
+end
+
+"""
+step! for non-PyArray state
+"""
+function Reinforce.step!(env::GymEnv{T}, a) where T
```

Comment:
I'm not a fan of copying code,
in the case of `Reinforce.step!`, please make them as a generated function.

Although, `Reinforce.reset!` is copied also, they are small enough for maintaining.

Reply:
> I'm not a fan of copying code
Yeah me neither, was just the simplest way to get this up and running. I will see what I can do to reduce the repetition, a Generated Function might be a good idea.

Reply:
I did it good, `step!` and `reset!` now have just one method each